src/bin/nydusd/fs_cache.rs

```diff
@@ -743,6 +747,166 @@ impl FsCacheHandler {
+        val * 0x61C88647
+    }
+
+    #[inline]
+    fn rol32(&self, word: u32, shift: i32) -> u32 {
+        word << (shift & 31) | (word >> ((-shift) & 31))
+    }
+
+    #[inline]
+    fn round_up_u32(&self, size: usize) -> usize {
+        (size + 3) / 4 * 4
+    }
+
+    //address from kernel fscache_hash()
+    fn fscache_hash(&self, salt: u32, data: &[u8]) -> u32 {
+        assert_eq!(data.len() % 4, 0);
+
+        let mut x = 0;
+        let mut y = salt;
+        let mut buf_le32: [u8; 4] = [0; 4];
+        let n = data.len() / 4;
+
+        for i in 0..n {
+            buf_le32.clone_from_slice(&data[i * 4..i * 4 + 4]);
+            let a = unsafe { std::mem::transmute::<[u8; 4], u32>(buf_le32) }.to_le();
+            x ^= a;
+            y ^= x;
+            x = self.rol32(x, 7);
+            x += y;
+            y = self.rol32(y, 20);
+            y *= 9;
+        }
+        self.hash_32(y ^ self.hash_32(x))
+    }
+
+    fn generate_cookie_path(
+        &self,
+        volume_path: &Path,
+        volume_key: &str,
+        cookie_key: &str,
+    ) -> (PathBuf, String) {
+        //calc volume hash
+        let mut volume_hash_key: Vec<u8> =
+            Vec::with_capacity(self.round_up_u32(volume_key.len() + 2));
+        volume_hash_key.push(volume_key.len() as u8);
+        volume_hash_key.append(&mut volume_key.as_bytes().to_vec());
+        while volume_hash_key.len() != volume_hash_key.capacity() {
+            volume_hash_key.push(0);
+        }
+        let volume_hash = self.fscache_hash(0, volume_hash_key.as_slice());
+
+        //calc cookie hash
+        let mut cookie_hash_key: Vec<u8> = Vec::with_capacity(self.round_up_u32(cookie_key.len()));
+        cookie_hash_key.append(&mut cookie_key.as_bytes().to_vec());
+        while cookie_hash_key.len() != cookie_hash_key.capacity() {
+            cookie_hash_key.push(0);
+        }
+        let dir_hash = self.fscache_hash(volume_hash, cookie_hash_key.as_slice());
+        let dir = format!("@{:02x}", dir_hash as u8);
+        let cookie = format!("D{}", cookie_key);
+        (volume_path.join(dir), cookie)
+    }
+
+    fn inuse(&self, cookie_dir: &Path, cookie_name: &str) -> Result<bool> {
+        env::set_current_dir(&cookie_dir)?;
+        let msg = format!("inuse {}", cookie_name);
+        let ret = unsafe {
+            libc::write(
+                self.file.as_raw_fd(),
+                msg.as_bytes().as_ptr() as *const u8 as *const libc::c_void,
+                msg.len(),
+            )
+        };
+        if ret < 0 {
+            let err = Error::last_os_error();
+            if let Some(e) = err.raw_os_error() {
+                if e == libc::EBUSY {
+                    return Ok(true);
+                }
+            }
+            Err(err)
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn cull(&self, cookie_dir: &Path, cookie_name: &str) -> Result<()> {
+        env::set_current_dir(&cookie_dir)?;
```

Comment:
It seems a little dangerous to change current working directory, should we start a working thread to do this?